plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ServerEvaluationStrategyProvider.java

```
@@ -0,0 +1,60 @@
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat Inc. - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.che.plugin.docker.machine;
+
+import java.util.Map;
+
+import org.eclipse.che.plugin.docker.machine.local.LocalDockerModule;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+import com.google.inject.name.Named;
+
+/**
+ * Provides {@code ServerEvaluationStrategies}. Intended to be injected into {@link DockerInstanceRuntimeInfo}
+ * instances to aid in {@link DockerInstanceRuntimeInfo#getServers()}
+ *
+ * <p>Strategy to be used is selected by property {@code che.docker.server_evaluation_strategy}, which should
+ * refer to a strategy bound in {@link LocalDockerModule}. If the strategy cannot be found, the default will
+ * be used.
+ *
+ * @author Angel Misevski <amisevsk@redhat.com>
+ * @see ServerEvaluationStrategy
+ */
+public class ServerEvaluationStrategyProvider implements Provider<ServerEvaluationStrategy> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ServerEvaluationStrategyProvider.class);
+
+    private static final String DEFAULT_STRATEGY_NAME = "default";
+
+    private ServerEvaluationStrategy strategy;
+
+    @Inject
+    public ServerEvaluationStrategyProvider(Map<String, ServerEvaluationStrategy> strategies,
+                                            @Named("che.docker.server_evaluation_strategy") String chosenStrategy) {
+        if (strategies.containsKey(chosenStrategy)) {
+            this.strategy = strategies.get(chosenStrategy);
+        } else {
+            LOG.warn(String.format("Property che.docker.server_evaluation_strategy=%s"
```

It will be hard to debug and realize what happened if developer made a typo in strategy name for example. So it is better to fail component start in that case instead of usage of default strategy.

---

That makes a lot of sense.

What is the preferred way of doing this in Che? I've changed it to throw a ServerException with a message now, but the error that appears while launching a workspace includes errors about being unable to inject the constructor, which is not very clean. Since I'm implementing `Provider<ServerEvaluationStrategy>`, we can't throw an exception in the get() method. I suppose implementing Provider is not strictly required, WDYT?

---

No, it can be provider de-facto but doesn't implement any interface

---

The further issue that MachineRuntimeInfo (and thus DockerInstanceRuntimeInfo) do not leave space for throwing a ServerException. Is there an internal way to signal an error otherwise? 

Alternatively, we could add throws declarations to MachineRuntimeInfo, and adjust as necessary, but DockerInstanceRuntimeInfo is used pretty widely in the codebase.

---

I think that declaring throwing of an exception in MachineRuntimeInfo is better to omit because, as you said, it is used in a lot of places and bothering all that code with catching exception would make a lot of code much more complicated.

So I think we need to find another way to achieve what you want. 

Can you elaborate on what problem you have faced with lack of exception throwing?

---

For useability, it would be ideal if typing the property incorrectly would cause workspaces to fail to start, displaying a message saying something along the lines of "Server Evaluation Strategy \<property\> does not correspond to any implemented strategy". The user could read their typo, and make the appropriate fix in che.env without too much trouble.

I'm not familiar enough with the relevant sections of the code to see an easy way to fail workspace starting and also display this message to the user without throwing an exception. This works relatively well, except that the only place I can throw this exception is in the ServerEvaluationStrategyProvider constructor, which litters the error message with "failed to inject constructor" and other information about the internals of Che. 

My initial approach was to add a error message to the log and use the default strategy, but I agree that this can easily lead to confusion. Another option is to log an error and use a no-op implementation of ServerEvaluationStrategy -- this would cause launching a workspace to fail, but the user would still have to view the logs to find their error.

How would you like me to surface this error to the user?

---

If you set this class to be eager singleton Guice will have to create it on Che bootstrap and mistake in strategy name would prevent successful start of Che. Maybe this variant is ok for you? 
In addition to that check and error throwing can be moved to **@PostConstruct** method. Then maybe message of Che start fail would be more satisfactory.

---

After playing around with it for awhile, I think I've got something that works. 

The Exception is thrown in a @PostConstruct method, which causes workspace startup to fail and display a user-readable message. Binding ServerEvaluationStrategyProvider as a Singleton caused an issue where Che would still start without displaying errors (except for in the logs), but be in an inconsistent state and have to be stopped from the docker cli directly.